src/libraries/System.Text.Json/src/Resources/Strings.resx

```diff
@@ -449,23 +449,29 @@
 1. {0}
 2. {1}</value>
   </data>
-  <data name="MetadataPreservedArrayInvalidProperty" xml:space="preserve">
-    <value>Invalid property '{0}' found within a JSON object that must only contain metadata properties and the nested JSON array to be preserved.</value>
+  <data name="MetadataInvalidPropertyInArrayMetadata" xml:space="preserve">
+    <value>A JSON object containing metadata for a nested array includes a non-metadata property '{0}'.</value>
   </data>
-  <data name="MetadataPreservedArrayPropertyNotFound" xml:space="preserve">
-    <value>One or more metadata properties, such as '$id' and '$values', were not found within a JSON object that must only contain metadata properties and the nested JSON array to be preserved.</value>
+  <data name="MetadataStandaloneValuesProperty" xml:space="preserve">
+    <value>A '$values' metadata property must always be preceded by other metadata properties, such as '$id' or '$type'.</value>
   </data>
   <data name="MetadataReferenceCannotContainOtherProperties" xml:space="preserve">
     <value>A JSON object that contains a '$ref' metadata property must not contain any other properties.</value>
   </data>
   <data name="MetadataReferenceNotFound" xml:space="preserve">
-    <value>Reference '{0}' not found.</value>
+    <value>Reference '{0}' was not found.</value>
   </data>
   <data name="MetadataValueWasNotString" xml:space="preserve">
-    <value>The '$id' and '$ref' metadata properties must be JSON strings. Current token type is '{0}'.</value>
+    <value>The '$id', '$ref' or '$type' metadata properties must be JSON strings. Current token type is '{0}'.</value>
   </data>
   <data name="MetadataInvalidPropertyWithLeadingDollarSign" xml:space="preserve">
-    <value>Properties that start with '$' are not allowed on preserve mode, either escape the character or turn off preserve references by setting ReferenceHandler to null.</value>
+    <value>Properties that start with '$' are not allowed in types that support metadata. Either escape the character or disable reference preservation and polymorphic deserialization.</value>
+  </data>
+  <data name="MetadataUnexpectedProperty" xml:space="preserve">
+    <value>The metadata property is either not supported by the type or is not the first property in the deserialized JSON object.</value>
+  </data>
+  <data name="MetadataDuplicateTypeProperty" xml:space="preserve">
+    <value>Deserialized object contains a duplicate '$type' metadata property.</value>
```

Comment:
Good question, I used `$type` as an alias for "type discriminator" in this context, expecting users applying custom type discriminators are advanced enough to understand this. In hindsight it might make more sense to parameterize the error message.